iiod/iiod.c

```diff
@@ -106,57 +110,255 @@ static const char *options_descriptions[] = {
 };
 
 #ifdef HAVE_AVAHI
-static AvahiSimplePoll *avahi_poll;
-static AvahiClient *avahi_client;
+
+/***
+ * Parts of the avahi code are borrowed from the client-publish-service.c
+ * https://www.avahi.org/doxygen/html/client-publish-service_8c-example.html
+ * which is an example in the avahi library. Copyright Lennart Poettering
+ * released under the LGPL 2.1 or (at your option) any later version.
+ */
+
+/* Global Data */
+
+static struct avahi_data {
+	AvahiThreadedPoll *poll;
+	AvahiClient *client;
+	AvahiEntryGroup *group;
+	char * name;
+} avahi;
+
+static void create_services(AvahiClient *c);
+static AvahiClient * client_new(void);
+
+static void client_free()
+{
+	/* This also frees the entry group, if any. */
+	if (avahi.client) {
+		avahi_client_free(avahi.client);
+		avahi.client = NULL;
+		avahi.group = NULL;
+	}
+}
+
+static void shutdown_avahi()
+{
+	/* Stop the avahi client, if it's running. */
+	if (avahi.poll)
+		avahi_threaded_poll_stop(avahi.poll);
+
+	/* Clean up the avahi objects. The order is significant. */
+	client_free();
+	if (avahi.poll) {
+		avahi_threaded_poll_free(avahi.poll);
+		avahi.poll = NULL;
+	}
+	if (avahi.name) {
+		INFO("Avahi: Removing service '%s'\n", avahi.name);
+		avahi_free(avahi.name);
+		avahi.name = NULL;
+	}
+}
 
 static void __avahi_group_cb(AvahiEntryGroup *group,
 		AvahiEntryGroupState state, void *d)
 {
+	/* Called whenever the entry group state changes */
+	if (!group)  {
+		ERROR("__avahi_group_cb with no valid group\n");
+		return;
+	}
+
+	avahi.group = group;
+
+	switch (state) {
+		case AVAHI_ENTRY_GROUP_ESTABLISHED :
+			INFO("Avahi: Service '%s' successfully established.\n",
+					avahi.name);
+			break;
+		case AVAHI_ENTRY_GROUP_COLLISION : {
+			char *n;
+			/* A service name collision, pick a new name */
+			n = avahi_alternative_service_name(avahi.name);
+			avahi_free(avahi.name);
+			avahi.name = n;
+			INFO("Avahi: Group Service name collision, renaming service to '%s'\n",
+					avahi.name);
+			create_services(avahi_entry_group_get_client(group));
```

Comment:
curios: is there a chance that services need to be free'd here before create_services()?

Reply:
no, when necessary - group_reset() is called. this does what avahi wants.